# Algorithmes gloutons - EXERCICES

Un algorithme glouton permet d'apporter une solution à un problème d'optimisation (*maximiser* ou *minimiser* une grandeur) tout en respectant certaines *contraintes*.


## Exercice 1

On cherche à sélectionner cinq nombres de la liste suivante en cherchant à avoir leur somme la plus grande possible (maximiser une grandeur) et en s'interdisant de choisir deux nombres voisins (contrainte).

| <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> | <!-- --> |
|---| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 15 | 4 | 20 |	17 | 11 | 8 | 11 | 16 |	7 | 14 | 2 | 7 | 5 | 17 | 19 | 18 |	4 |	5 |	13 | 8 |

Comme on souhaite avoir le plus grand résultat final, la stratégie gloutonne consiste à choisir à chaque étape le plus grand nombre possible dans les choix restants.

1. Appliquez cet algorithme glouton sur le tableau.
2. Vérifiez que $\{20, 18, 17, 16, 15\}$ est une autre solution possible.
3. Que dire de la solution gloutonne ?

## Exercice 2 : le problème du voyageur

Un voyageur a ciblé plusieurs villes qu'il souhaite visiter. Il cherche un itinéraire passant par toutes ces villes et qui minimise la distance totale parcourue. Les villes peuvent être visitées dans n'importe quel ordre mais aucune ne doit être négligée, et le visiteur doit revenir à la fin à sa ville de départ.

Le voyageur part de Nancy et souhaite visiter Metz, Paris, Reims et Troyes, avant de retourner à Nancy.

Voici un tableau donnant les distances kilométriques entre chacune des ces villes.

![tableau des distances](data/tableau_voyageur.png)

1. Quelle est la stratégie gloutonne à mettre en oeuvre ?
2. Mettez en oeuvre cette stratégie et donnez la solution.
3. Calculez la distance totale pour le parcours Metz - Reims - Paris - Troyes (départ et arrivée à Nancy sous-entendus)
4. Que dire de la solution gloutonne ?

## Exercice 3 : le problème du sac à dos

On rappelle qu'il y a plusieurs stratégies gloutonnes pour donner une solution à ce problème

- **Stratégie 1** : prendre toujours l'objet de plus grande valeur n'excédant pas la capacité restante (il faut trier préalablement par valeur décroissante)
- **Stratégie 2** : prendre toujours l'objet de plus faible masse (il faut trier préalablement par masse croissante)
- **Stratégie 3** : prendre toujours l'objet de plus grand rapport $\frac{\text{valeur}}{\text{masse}}$ n'excédant pas la capacité restante (il faut trier préalablement par rapport $\frac{\text{valeur}}{\text{masse}}$ décroissant)

### **Exemple 1**

Considérons les objets suivants et un sac de capacité maximale 2 kg.

| objet  |  A  |  B  |  C  |
|:------:|:---:|:---:|:---:|
|  masse (kg) |  1,5 |  2 |  0,3  |
| valeur (€) | 200 | 500 | 400 |
|  $\frac{\text{valeur}}{\text{masse}}$ | 133,33... | 250 | 1333,33... |

1. Appliquez chacune des stratégies à ce problème.
2. Quelle est la meilleure stratégie dans ce cas ?
3. Listez toutes les combinaisons possibles, puis celles respectant la contrainte et déduisez-en la solution optimale au problème. Comparez-la aux stratégies gloutonnes.

### **Exemple 2** 

Même questions mais avec un sac de capacité maximale 3,5 kg.


## **Exemple 3**

Considérons les objets suivants et un sac de capacité maximale 2 kg.

| objet  |  A  |  B  |  C  |
|:------:|:---:|:---:|:---:|
|  masse (kg) |  1,5 |  2 |  0,3  |
| valeur (€) | 200 | 500 | 400 |
|  $\frac{\text{valeur}}{\text{masse}}$ | 133,33... | 250 | 1333,33... |

Correction : Ex 3 qu

- la stratégie 1 (choix par valeur décroissante) donne le sac $\{B\}$ d'une valeur de 500 € ; 
- la stratégie 2 (choix par masse croissante) donne le sac $\{C, A\}$ d'une valeur de 600 € ;
- la stratégie 3 (choix par rapport valeur/masse décroissante) donne le sac $\{C, A\}$ d'une valeur de 600 €.

Les stratégies 2 et 3 sont meilleures et donnent dans ce cas la solution optimale.



- la stratégie 1 (choix par valeur décroissante) donne le sac $\{B, C\}$ d'une valeur de 900 € ; 
- la stratégie 2 (choix par masse croissante) donne le sac $\{C, A\}$ d'une valeur de 600 € ;
- la stratégie 3 (choix par rapport valeur/masse décroissante) donne le sac $\{C, B\}$ d'une valeur de 900 €.

Cette fois-ci, ce sont les stratégies 1 et 3 qui sont meilleures et donnent dans ce cas la solution optimale.


## Exercice 4

>**Objectifs** : 
- implémenter un algorithme glouton pour trouver une solution à un problème
- implémenter l'algorithme de *force brute* pour déterminer la solution optimale et la comparer à la solution gloutonne

### Introduction

Nous disposons d’une clé USB qui est déjà bien remplie et sur laquelle il ne reste que 8 Go de libre. Nous souhaitons copier sur cette  clé  des  fichiers  vidéos  pour  l’emporter  en  voyage.  Chaque  fichier  a  un poids  et  chaque  vidéo  a  une  durée.  La  durée  n’est pas proportionnelle à la taille car les fichiers sont de format différents, certaines vidéos sont de grande qualité, d’autres sont très compressées. Le tableau qui suit présente les 10 fichiers disponibles avec les durées données en minutes.

| Nom 	| Durée en min (valeur) 	| Poids |
|---------	|-------	|---------	|
| Vidéo A 	| 114   	| 4.5 Go 	|
| Vidéo B 	| 85    	| 630 Mo  	|
| Vidéo C 	| 40    	| 3.35 Go 	|
| Vidéo D 	| 4     	| 85 Mo   	|
| Vidéo E 	| 18    	| 2,15 Go 	|
| Vidéo F 	| 80    	| 2,71 Go 	|
| Vidéo G 	| 5     	| 320 Mo  	|
| Vidéo H 	| 86    	| 3.7 Go  	|
| Vidéo I 	| 64     	| 2.4 Go  	|
| Vidéo J 	| 12     	| 6.4 Go  	|

>**Problème** : Quelles vidéos copier sur la clé USB pour que la durée des vidéos soient la plus grande possible tout en ne dépassant pas 8 Go ?

**Question 1** : Quelle est la valeur que l'on cherche à maximiser/minimiser ? Quelle est la contrainte ?

**Question 2** : Quel problème reconnaissez-vous ici ?

>**Notation** : Dans la suite, on identifiera la *durée* d'une vidéo à sa *valeur* pour garder le même vocabulaire que dans le cours.

Vous allez implémenter (= programmer) une solution gloutonne pour résoudre ce problème, puis vous implémenterez l'algorithme de *force brute* qui donnera la solution optimale au problème. La stratégie gloutonne retenue sera de toujours prendre la vidéo de plus grand rapport $\frac{\text{valeur}}{\text{poids}}$ n'excédant pas la capacité restante.

### Représentation des données

On utilise un dictionnaire pour représenter une vidéo, la valeur étant la durée en minute et le poids étant donné en Go.

```python
{'nom': 'Vidéo A', 'valeur': 114, 'poids': 4.57}
```

On peut alors mémoriser les 10 vidéos dans le tableau `table_videos` suivant.

In [27]:
table_videos = [{'nom' : 'video A', 'valeur' : 114, 'poids' : 4.5},
                {'nom' : 'video B', 'valeur' : 85, 'poids' : 0.63},
                {'nom' : 'video C', 'valeur' : 40, 'poids' : 3.35},
                {'nom' : 'video D', 'valeur' : 4, 'poids' : 0.085},
                {'nom' : 'video E', 'valeur' : 18, 'poids' : 2.15},
                {'nom' : 'video F', 'valeur' : 80, 'poids' : 2.71},
                {'nom' : 'video G', 'valeur' : 5, 'poids' : 0.32},
                {'nom' : 'video H', 'valeur' : 74, 'poids' : 3.7},
                {'nom' : 'video I', 'valeur' : 64, 'poids' : 2.4},
                {'nom' : 'video J', 'valeur' : 12, 'poids' : 6.4}]

On peut alors accéder aux éléments de cette table.

In [28]:
table_videos[0]['nom']

'video A'

In [29]:
table_videos[3]['valeur']

4

In [30]:
table_videos[9]['poids']

6.4

### Ecriture des fonctions utiles

#### Accès au poids d'une vidéo

Pour résoudre le problème, on a besoin d'accéder au poids des vidéos pour vérifier la contrainte du poids maximal. La fonction suivante permet de renvoyer le poids d'une vidéo entrée en paramètre (sous la forme d'un dictionnaire vu précédemment). Quelques assertions devant être vérifiées par la fonction ont été écrites.

In [31]:
def poids(video):
    return video['poids']

assert poids({'nom' : 'video A', 'valeur' : 114, 'poids' : 4.5}) == 4.5
assert poids({'nom' : 'video D', 'valeur' : 4, 'poids' : 0.085}) == 0.085
assert poids({'nom' : 'video J', 'valeur' : 12, 'poids' : 6.4}) == 6.4

#### Accès à la durée d'une vidéo

De même, pour connaître la durée totale d'une solution, il est nécessaire de pouvoir accéder aux durées de chacune des vidéos retenues.

**Question 2** : Ecrivez une fonction `duree(video)` qui renvoie la durée (= valeur) d'une video entrée en paramètre (sous la forme d'un dictionnaire vu précédemment). Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

In [32]:
def duree(video):
    # code à écrire ici
    return video['valeur']

In [33]:
assert duree({'nom' : 'video A', 'valeur' : 114, 'poids' : 4.5}) == 114
assert duree({'nom' : 'video D', 'valeur' : 4, 'poids' : 0.085}) == 4
assert duree({'nom' : 'video J', 'valeur' : 12, 'poids' : 6.4}) == 12

#### Accès au rapport $\frac{\text{valeur}}{\text{poids}}$

De même, la stratégie gloutonne choisie prévoit de prendre la vidéo de plus grand rapport $\frac{\text{valeur}}{\text{poids}}$ n'excédant pas la capacité restante. Il est donc nécessaire d'accéder à cette donnée pour chaque vidéo.

**Question 3** : Ecrivez une fonction `rapport_valeur_poids(video)` qui prend en paramètre une vidéo et renvoie son rapport $\frac{\text{valeur}}{\text{poids}}$. Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

In [34]:
def rapport_valeur_poids(video):
    # code à écrire ici
    return duree(video)/poids(video)  # ou video['valeur']/video['poids']

In [35]:
assert rapport_valeur_poids({'nom' : 'video H', 'valeur' : 74, 'poids' : 3.7}) ==  20
assert rapport_valeur_poids({'nom' : 'video J', 'valeur' : 12, 'poids' : 6.4}) == 1.875
# vous remarquerez que l'on a pris soin de sélectionner les vidéos dont le rapport peut être représenté de manière exacte en machine

Enfin, comme le choix se fait par rapport $\frac{\text{valeur}}{\text{poids}}$ décroissant, il est judicieux de trier au préalable les vidéos par ordre décroissant selon ce rapport.

**Question 4** : Ecrivez une fonction `tri_rapport_decroissant(table_videos)` qui renvoie un *nouveau tableau* dans lequel les fichiers ont été triés par rapport $\frac{\text{valeur}}{\text{poids}}$ décroissant. Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

*Indications* : l'instruction `help(sorted)` vous permet d'afficher l'aide de la fonction `sorted`. Si besoin, revoir le notebook sur le tri d'une table (Séquence 5, Chapitre 1, Activité 2).

In [36]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [99]:
def tri_rapport_decroissant(table_videos):
    table_triee = sorted(table_videos, key = rapport_valeur_poids, reverse = True)
    return table_triee

def tri_valeur_decroissante(tables_videos):
    table_triee = sorted(table_videos, key = duree, reverse = True)
    return table_triee

def tri_poids_croissant(tables_videos):
    table_triee = sorted(table_videos, key = poids)
    return table_triee

In [100]:
assert tri_rapport_decroissant(table_videos) == [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
                                                 {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
                                                 {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
                                                 {'nom': 'video I', 'valeur': 64, 'poids': 2.4},
                                                 {'nom': 'video A', 'valeur': 114, 'poids': 4.5},
                                                 {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
                                                 {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
                                                 {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
                                                 {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
                                                 {'nom': 'video J', 'valeur': 12, 'poids': 6.4}]

#### Calculer le poids total et la durée totale d'une table de vidéos

Enfin, il sera nécessaire de pouvoir calculer le poids et la durée totale d'une table de vidéos pour savoir si la contrainte de capacité maximale est respectée d'une part, et pour connaître la durée totale qui est la valeur à maximiser.

**Question 5** : Ecrivez une fonction `poids_total(table_videos)` qui renvoie le poids total des vidéos présentes dans `table_videos`. Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

In [41]:
def poids_total(table_videos):
    poids_t = 0
    for video in table_videos:
        poids_t = poids_t + video['poids']
    return poids_t

In [43]:
table1 = [{'nom': 'video J', 'poids': 6.4, 'valeur': 12}]
table2 = [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
           {'nom': 'video J', 'poids': 6.4, 'valeur': 12}]
table3 = [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
          {'nom': 'video B', 'poids': 0.63, 'valeur': 85},
          {'nom': 'video D', 'poids': 0.085, 'valeur': 4}]

assert poids_total(table1) == 6.4   # on compare des nombres réels mais on a pris soin de bien choisir les exemples
assert poids_total(table2) == 10.9
assert poids_total(table3) == 5.215
assert poids_total([]) == 0   # table vide testée

**Question 6** : Ecrivez une fonction `duree_totale(table_videos)` qui renvoie la durée totale des vidéos présentes dans `table_videos`. Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

In [44]:
def duree_totale(table_videos):
    duree_t = 0
    for video in table_videos:
        duree_t = duree_t + video['valeur']
    return duree_t

In [45]:
table1 = [{'nom': 'video J', 'poids': 6.4, 'valeur': 12}]
table2 = [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
           {'nom': 'video J', 'poids': 6.4, 'valeur': 12}]
table3 = [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
          {'nom': 'video B', 'poids': 0.63, 'valeur': 85},
          {'nom': 'video D', 'poids': 0.085, 'valeur': 4}]

assert duree_totale(table1) == 12
assert duree_totale(table2) == 126
assert duree_totale(table3) == 203
assert duree_totale([]) == 0   # table vide testée
assert duree_totale(table_videos) == 496 # table avec toutes les vidéos

### Implémentation de l'algorithme glouton

**Question 7** : en vous inspirant de l'implémentation de l'algorithme glouton du rendu de monnaie (voir cours), écrivez une fonction `glouton(table_videos, poids_max)` qui prend en paramètre une table de vidéos `table_videos` et un poids maximal `poids_max` et renvoie une table `solution_gloutonne` contenant la solution gloutonne au problème. La solution gloutonne est donnée dans l'assertion qui suit.

In [103]:
def glouton(table_videos, poids_max):
    table_triee = tri_poids_croissant(table_videos)
    poids_total = 0
    solution_gloutonne = []
    # à compléter
    i = 0
    while poids_total < poids_max and i < len(table_triee):  # tant que le poids max n'est pas atteint et que l'on n'a pas traité toutes les vidéos
        video = table_triee[i]
        poids_video = poids(video)
        if poids_total + poids_video <= poids_max:  # on ajoute la vidéo d'indice i à notre solution
            solution_gloutonne.append(video)
            poids_total = poids_total + poids_video  # on met à jour le poids total de note solution
        i = i + 1
    return solution_gloutonne

In [104]:
glouton(table_videos, 8)

[{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
 {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
 {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
 {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
 {'nom': 'video I', 'valeur': 64, 'poids': 2.4}]

In [58]:
assert glouton(table_videos, 8) == [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
                                    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
                                    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
                                    {'nom': 'video I', 'valeur': 64, 'poids': 2.4},
                                    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}]

**Question 8** : Quel est le poids total de la solution gloutonne ? Quelle est sa durée totale ?

In [65]:
# à compléter
solution_gloutonne = glouton(table_videos, 8)
print("poids total :", poids_total(solution_gloutonne), "Go")
print("durée totale :", duree_totale(solution_gloutonne), "min")

poids total : 6.145 Go
durée totale : 238 min


### Implémentation de la *force brute*

Le principe est simple : il faut étudier tous les cas possibles. Ainsi, pour appliquer cette stratégie, il faut :
1. d'abord *énumérer* toutes les combinaisons possibles de vidéos 
2. puis conserver celles dont la capacité maximale n'est pas dépassée
3. enfin, trouver la meilleure solution parmi les combinaisons restantes

#### Etape 1 : énumération de toutes les combinaisons

Il s'agit de la difficulté majeure. Dans un combinaison de vidéos, chaque vidéo de départ est prise ou non, il s'agit d'une donnée binaire. Ainsi, une approche consiste à créer des mots binaires représentant chaque combinaison. Par exemple, le mot '1101000001' signifie qu'on prend les vidéos A, B, D et J tandis que le mot '1111111111' signifie que l'on prend toutes les vidéos.

Dans notre exemple, nous avons 10 vidéos donc il y a $2^{10} = 1024$ combinaisons possibles. De manière générale il y a donc $2^n$ combinaisons pour un ensemble de $n$ vidéos.

Pour obtenir notre liste de mots binaires, on peut parcourir tous les entiers compris entre $0$ et $2^n-1$, les convertir en binaire et les ajouter à une liste appelée `combinaisons`.

>**Astuce** : la fonction `bin` prend un entier en paramètre et renvoie sa valeur binaire sous forme d'une chaîne de caractères. Par exemple, `bin(12)` renvoie la chaîne `'0b1101'`. Il suffira de supprimer les caractères `'0b'` en tête pour obtenir l'écriture binaire. On veillera également à compléter avec des zéros devant pour obtenir un mot de la longueur désirée.

La fonction `table_combinaisons` suivante prend en paramètre une table de vidéos et renvoie une liste `combinaisons` contenant tous les mots binaires correspondant aux différentes combinaisons possibles de vidéos dans `table_videos`.

In [66]:
def table_combinaisons(table_videos):
    n = len(table_videos)  # nombre de vidéos
    table_entiers = [i for i in range(2**n)]  # création d'un tableau avec tous les entiers entre 0 et 2**n-1
    #print(table_entiers)
    table_binaire = [bin(i)[2:] for i in table_entiers]  # conversion binaire des entiers du tableau précédent, le [2:] permet de supprimer les caractères de tête '0b' renvoyés par la fonction bin
    #print(table_binaire)
    combinaisons = ['0'*(n-len(k)) + k for k in table_binaire]  # ajout des zéros pour obtenir des mots binaires de longueur n
    return combinaisons

On peut vérifier que sur notre exemple, cette fonction renvoie bien un tableau avec tous les mots binaires de longueur 10. Je vous encourage à décommenter les lignes avec les affichages pour voir le contenu de chaque tableau.

In [67]:
table_combinaisons(table_videos)

['0000000000',
 '0000000001',
 '0000000010',
 '0000000011',
 '0000000100',
 '0000000101',
 '0000000110',
 '0000000111',
 '0000001000',
 '0000001001',
 '0000001010',
 '0000001011',
 '0000001100',
 '0000001101',
 '0000001110',
 '0000001111',
 '0000010000',
 '0000010001',
 '0000010010',
 '0000010011',
 '0000010100',
 '0000010101',
 '0000010110',
 '0000010111',
 '0000011000',
 '0000011001',
 '0000011010',
 '0000011011',
 '0000011100',
 '0000011101',
 '0000011110',
 '0000011111',
 '0000100000',
 '0000100001',
 '0000100010',
 '0000100011',
 '0000100100',
 '0000100101',
 '0000100110',
 '0000100111',
 '0000101000',
 '0000101001',
 '0000101010',
 '0000101011',
 '0000101100',
 '0000101101',
 '0000101110',
 '0000101111',
 '0000110000',
 '0000110001',
 '0000110010',
 '0000110011',
 '0000110100',
 '0000110101',
 '0000110110',
 '0000110111',
 '0000111000',
 '0000111001',
 '0000111010',
 '0000111011',
 '0000111100',
 '0000111101',
 '0000111110',
 '0000111111',
 '0001000000',
 '0001000001',
 '00010000

#### Etape 2 : conservation des combinaisons possibles

On cherche maintenant à conserver uniquement les combinaisons ne dépassant pas la capacité maximale de 8 Go. Par exemple : 
- `'0000000001'` est à conserver puisqu'il s'agit uniquement de la vidéo J de 6.4 Go
- `'1000000001'` n'est pas à conserver car le poids total de cette combinaison (vidéos A et J) vaut 4.5 + 6.4 = 10.9 Go, ce qui dépasse la capacité maximale.

Pour cela, il faut d'abord associer à chaque combinaison la liste de vidéos correspondantes.

**Question 9** : Ecrivez une fonction `construction_table(combi)` qui prend en paramètre une combinaison `combi` (mot binaire écrit sous forme d'une chaîne de caractères) et qui renvoie une table `table_correspondante` avec les vidéos correspondantes de `table_videos`. Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

In [70]:
def construction_table(combi, table_videos):
    table_correspondante = []
    # à compléter
    for i in range(len(combi)) :  # i parcourt les indices des caractères de la chaîne combi
        if combi[i] == '1':   # si c'est un 1
            table_correspondante.append(table_videos[i])  # on ajoute la vidéo d'indice i à table_correspondante
    return table_correspondante

In [71]:
assert construction_table('0000000001', table_videos) == [{'nom': 'video J', 'poids': 6.4, 'valeur': 12}]
assert construction_table('1000000001', table_videos) == [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
                                                          {'nom': 'video J', 'poids': 6.4, 'valeur': 12}]
assert construction_table('1101000000', table_videos) == [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
                                                          {'nom': 'video B', 'poids': 0.63, 'valeur': 85},
                                                          {'nom': 'video D', 'poids': 0.085, 'valeur': 4}]

Maintenant que l'on sait construire la table de vidéos correspondant à une combinaison, on peut utiliser la fonction `poids_total` pour savoir si la table respecte ou non la capacité maximale autorisée.

**Question 10** : Ecrivez une fonction `extraire_tables_valides(table_videos, poids_max)` qui renvoie une liste `tables_valides` contenant les combinaisons de vidéos de `table_videos` respectant la contrainte `poids_max` de poids maximal. Quelques assertions devant être vérifiées par votre fonction sont données ci-dessous.

In [92]:
def extraire_tables_valides(table_videos, poids_max):
    tables_valides = []
    # à compléter
    combinaisons = table_combinaisons(table_videos) # construction de toutes les combinaisons binaires
    for combi in combinaisons:  # pour chaque combinaison binaire
        table_correspondante = construction_table(combi, table_videos)  # construction de la table correspondante
        if poids_total(table_correspondante) <= poids_max:  # si la contrainte de poids est respectée
            tables_valides.append(table_correspondante)  # on ajoute la table à la liste de tables valides
    return tables_valides

In [93]:
table2 = [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
           {'nom': 'video J', 'poids': 6.4, 'valeur': 12}]

assert extraire_tables_valides(table2, 4) == [[]]
assert extraire_tables_valides(table2, 5) == [[], 
                                                         [{'nom': 'video A', 'poids': 4.5, 'valeur': 114}]]
assert extraire_tables_valides(table2, 7) == [[],
                                                         [{'nom': 'video J', 'poids': 6.4, 'valeur': 12}],
                                                         [{'nom': 'video A', 'poids': 4.5, 'valeur': 114}]]
assert extraire_tables_valides(table2, 11) == [[],
                                                          [{'nom': 'video J', 'poids': 6.4, 'valeur': 12}],
                                                          [{'nom': 'video A', 'poids': 4.5, 'valeur': 114}],
                                                          [{'nom': 'video A', 'poids': 4.5, 'valeur': 114},
                                                           {'nom': 'video J', 'poids': 6.4, 'valeur': 12}]]

L'appel `extraire_tables_valides(table_videos, 8)` permet de renvoyer la liste de toutes les combinaisons de vidéos d'un poids total inférieur ou égal à 8 Go. Il y en a beaucoup, voici la liste de tables valides que l'on obtient (essayez !).

In [90]:
tables_valides = [
    [],
    [{'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video J', 'valeur': 12, 'poids': 6.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video H', 'valeur': 74, 'poids': 3.7}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video C', 'valeur': 40, 'poids': 3.35},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32},
    {'nom': 'video I', 'valeur': 64, 'poids': 2.4}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15}],
    [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
    {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
    {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
    {'nom': 'video E', 'valeur': 18, 'poids': 2.15},
    {'nom': 'video G', 'valeur': 5, 'poids': 0.32}]]

#### Etape 3 : trouver la meilleure solution

C'est très simple : parmi les combinaisons de vidéos valides, il suffit de trouver celle dont la durée est maximale. Cela revient à faire une recherche de maximum. On parcourt toutes les combinaisons, pour chaque combinaison, on calcule la durée totale et on met à jour le maximum si besoin.

**Question 11** : Ecrivez une fonction `solution_force_brute(table_video, poids_max)` qui renvoie la solution optimale au problème, c'est-à-dire la combinaison maximisant la durée totale des vidéos tout en respectant la contrainte maximale de poids.

In [97]:
def solution_force_brute(table_videos, poids_max):
    tables_valides = extraire_tables_valides(table_videos, poids_max) # extraction des tables valides
    # recherche de la durée maximale
    solution_optimale = tables_valides[0]  # initialisation à la première table
    duree_max = duree_totale(solution_optimale)
    for table in tables_valides:  # pour chaque table valide
        duree_table = duree_totale(table)  # on calcule sa durée totale
        if duree_table > duree_max:  # si celle-ci est plus grande que la durée max déjà trouvée
            solution_optimale = table  # cette table devient notre solution optimale
            duree_max = duree_table  # et sa durée est la nouvelle durée max de référence
    return solution_optimale, duree_max  # on renvoie la solution optimale et sa durée

In [98]:
solution_force_brute(table_videos, 8)

([{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
  {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
  {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
  {'nom': 'video F', 'valeur': 80, 'poids': 2.71}],
 283)

L'appel `solution_force_brute(table_videos, 8)` renvoie la solution optimale suivante d'une durée totale maximale égale à 283 minutes.

```python
solution_optimale = [{'nom': 'video A', 'valeur': 114, 'poids': 4.5},
                     {'nom': 'video B', 'valeur': 85, 'poids': 0.63},
                     {'nom': 'video D', 'valeur': 4, 'poids': 0.085},
                     {'nom': 'video F', 'valeur': 80, 'poids': 2.71}]
```

On constate donc que la solution gloutonne $\{B, D, F, I, G\}$ d'une durée totale de 238 min (poids total : 6.145 Go) n'est pas optimale puisque la force brute prouve que l'on peut obtenir une durée totale de 283 min avec les vidéos $\{A,B, D, F\}$ (poids total 7.925 Go).

>**Remarque** : il était possible d'appliquer la force brute car le nombre de vidéos de départ était petit (10), mais cela aurait été impossible pour un grand nombre de vidéos car le nombre de combinaisons à énumérer dépasserait très vite les capacités d'un ordinateur aussi puissant soit-il.

#### Etape 2 :  

>**Questions**
1. Appliquez chacune de ces stratégies gloutonnes à la situation précédente. 
2. Dans chaque cas, prouvez que la solution obtenue n'est pas optimale.

Dans le premier cas, en choisissant les objets par valeur décroissante, l'algorithme glouton donne le sac $\{A, D\}$ d'une valeur 4800 + 500 = 5300 €.

Dans le second cas, en choisissant les objets par masse croissante, l'algorithme glouton donne le sac $\{D, C, B\}$ d'une valeur 500 + 3000 + 4000 = 7500 €.

Pour le dernier cas, commençons par calculer le rapport $\frac{\text{valeur}}{\text{masse}}$ de chaque objet.

| objet  |  A  |  B  |  C  |  D  |
|:------:|:---:|:---:|:---:|:---:|
|  $\frac{\text{valeur}}{\text{masse}}$ |  600 |  800 |  750  |  500 |

Ainsi, en choisissant par rapport décroissant, l'algorithme glouton donne le sac $\{B, C, D\}$ d'une valeur 7500 €.